### ISSUE TITLE
How to train PSPNet/DeepLabV3+ just as the result in your paper

### ISSUE THREAD

**Issue by Lucky-Light-Sun** [hkchengrex/CascadePSP#70]
Hi, thanks for your great contributions to our community. 

I have some questions about the training details, especially about the PSPNet/DeepLabV3+ in your paper [CascadePSP: Toward Class-Agnostic and Very High-Resolution Segmentation via Global and Local Refinement](https://ieeexplore.ieee.org/document/9157093/) Table 3.  Are there any techniques to achieve such great results? 

I followed the reference code "https://github.com/Lextal/pspnet-pytorch" and trained on 224x224 images and test on 1024x1024 BigDataset as the dataset prepared strategy. But the best result for iou metric is about 45% (your pspnet without CRM module is 90.49%!).

So I wonder if there are some techniques for you to train the model. Can you explain it? Thank you very much! It will be the best if you can release the PSPNet/DeepLabV3+ codes and weights. 

Thansk again for your awesome work. Looking forward to your reply!

![image](https://github.com/hkchengrex/CascadePSP/assets/60919766/e37f878f-abea-4b99-ba07-cc4560fefd3d)

**Comment by hkchengrex** [comment]
1. We did not retrain. That is clear from the caption that you highlighted.
2. That reference is for our own implementation. Again, as mentioned in the caption, we used the official implementation for obtaining the baseline results: https://github.com/hszhao/PSPNet

**Comment by Lucky-Light-Sun** [comment]
> 1. We did not retrain. That is clear from the caption that you highlighted.
> 2. That reference is for our own implementation. Again, as mentioned in the caption, we used the official implementation for obtaining the baseline results: https://github.com/hszhao/PSPNet

Thanks for your reply. I understand what you mean, but there still some questions. The given weights (`ADE20K`, `VOC2012`, `cityscapes`) are multi-class prediction. For example, ADE20K dataset has 150 classes. How do you use the prediction result?
For example, I made a test for BIG evaluation dataset. And the pspnet prediction using ADE20K pretrained weight and dataset ground truth are shown below:
I want to use the cls 0 as the background cls, but the result below makes me believe that there maybe some better ways to get such great result in your paper.

![image](https://github.com/hkchengrex/CascadePSP/assets/60919766/7050e9e6-65c5-4fac-beb5-142dfeac018a)

**Comment by hkchengrex** [comment]
BIG is annotated with the same classes as in PASCAL. Just isolate the class that you need (in this case, car).

**Comment by Lucky-Light-Sun** [comment]
> BIG is annotated with the same classes as in PASCAL. Just isolate the class that you need (in this case, car).

Hi, I use two version PSPNet (101 & 50) with pretrained weight for PASCAL VOC 2012 and test it with two methods. The first way is to isolate the class just as you said before. The second way is to view all the classes except background class as the target label. Just as you said, the former way is great, and the iou reaches about 87%(using PSPNet101 pretrained on VOC dataset). But it is still 3% lower than the score in your paper. I wonder whether if you also use the same weight? Cause PSPNet50 pretrained in VOC reaches lower score in my way.
If the pretrained weight I used is the same as you used. I think there maybe something wrong about my inference process. So I visualize the inference-process, hope you can see it clearly.

I want to ask what's the config in your experiment? Such as the crop_size, stride size, mean and std? For mean and std I use the default config in VOC dataset. Looking forward to your reply!

![image](https://github.com/hkchengrex/CascadePSP/assets/60919766/e8c25b78-1e5d-46e6-8e24-292bfa94da50)

**Comment by hkchengrex** [comment]
You should not crop. See Section 4.3.

**Comment by Lucky-Light-Sun** [comment]
> You should not crop. See Section 4.3.

Hi, can you give me more details? I tried the method written in your paper(the photo below). 

Using the PSPNet101 weight pretrained on VOC. Then rescale the long axes to 512 which maintain the height-width ratio same. And then due to the PSPNet architecture, I pad the rescale width/height cause the model need the width/height satisfy `(height/weight - 1) % 8 == 0`. I use cv2.copyMakeBorder with param borderType=cv2.BORDER_CONSTANT to pad image.  

In order to normalize the tensor, I tried the default mean/std in PSPNet code and tried the default BIG dataset mean/std too. However the result is still no change in my experiment. The iou metric is abount 87%. 

So can you give me more information please?  For example, the pretrained weight that you used, the norm value mean/std and so on. Should I use ADE20k or cityscapes pretrained weight? 

Looking forward to your reply!

![image](https://github.com/hkchengrex/CascadePSP/assets/60919766/f1b63591-1413-48fd-9eab-4729bc0f265f)

**Comment by hkchengrex** [comment]
Sound right. No changes were made to std/mean. Make no sense to use ADE20K/cityscape. Are you using our script to compute IoU?

You can also compare with our provided results: https://github.com/hkchengrex/CascadePSP/blob/master/docs/dataset.md

**Comment by Lucky-Light-Sun** [comment]
Thanks for your advice! 

I test it and find the 'error image' (just as below plane). But I don't know why it happened. I guess the reason is that, your pspnet version may come from this "https://github.com/hszhao/PSPNet" which was implemented using `caffe`. But I'm not familiar with `caffe`. So I use the pytorch version "https://github.com/hszhao/semseg" which was implemented using `pytorch`. 
To be honest, since the two repoes were both implemented by the same auther, hszhao and both are official codes, maybe the pspnet weight is just the same. So I'm still not sure about the reason that caused lower iou in my code.

Do you have more ideas? Thanks for your advice again!

![image](https://github.com/hkchengrex/CascadePSP/assets/60919766/f92cfa51-1872-4cb4-b853-f007768a0956)

**Comment by hkchengrex** [comment]
I'm afraid that I can only provide limited help. My suggestion for you is to use our precomputed results.

**Comment by Lucky-Light-Sun** [comment]
> I'm afraid that I can only provide limited help. My suggestion for you is to use our precomputed results.

Hi, I wanna ask that which repo did you use before? The `caffe` version or `pytorch` version?

**Comment by hkchengrex** [comment]
The one that is listed as "source" here: https://github.com/hkchengrex/CascadePSP/blob/master/docs/dataset.md#segmentation-results

**Comment by Lucky-Light-Sun** [comment]
> The one that is listed as "source" here: https://github.com/hkchengrex/CascadePSP/blob/master/docs/dataset.md#segmentation-results

Thanks so much for your help. I have no question now!
